README.md
# NimbleStrftime

`nimble_strftime` is a simple and fast library for formatting datetimes into
strings based on the `strftime` tool found on UNIX-like systems.

## Examples

Once installed, you can format your calendar types right away:

```elixir
iex> datetime = ~U[2019-08-26 13:52:06.0Z]
# year(2 digits)-month-day hour(in a 12 hour clock):minute:second AM/PM
iex> NimbleStrftime.format(datetime, "%y-%m-%d %I:%M:%S %p")
"19-08-26 01:52:06 PM"

# day_of_week_abbreviated, month day_of_month year
iex> NimbleStrftime.format(datetime, "%a, %B %d %Y")
"mon, august 26 2019"

# preferred datetime, default setting "%Y-%m-%d %H:%M:%S"
iex> NimbleStrftime.format(datetime, "%c")
"2019-08-26 13:52:06"
```

You can also pass configuration parameters to set preferred formats,
set the size of abbreviated names and change the names of months,
week days, am and pm:

```elixir
iex> datetime = ~U[2019-08-26 13:52:06.0Z]

# preferred datetime, configured to something else
iex> NimbleStrftime.format(datetime, "%c", preferred_datetime: "%H:%M:%S %d-%m-%y")
"13:52:06 26-08-19"

# day_of_week configured to another language
iex> NimbleStrftime.format(
...>  datetime,
...>  "%A",
...>  day_of_week_names: fn day_of_week ->
...>    {"segunda-feira", "terça-feira", "quarta-feira", "quinta-feira",
...>    "sexta-feira", "sábado", "domingo"}
...>    |> elem(day_of_week - 1)
...>  end
...>)
"segunda-feira"

# month_name with settings for custom abbreviation and names
iex> NimbleStrftime.format(
...>  datetime,
...>  "%B",
...>  abbreviated_month_names: fn month ->
...>    {"янв", "февр", "март", "апр", "май", "июнь",
...>    "июль", "авг", "сент", "окт", "нояб", "дек"}
...>    |> elem(month - 1)
...>  end
...>)
# => "авг"
```

For more information, please consult the [online documentation](https://hexdocs.pm/nimble_strftime/NimbleStrftime.html)

## Installation

Add `nimble_strftime` to your dependencies:

```elixir
def deps do
  [
    {:nimble_strftime, "~> 0.1.0"}
  ]
end
```

## Nimble*

Other nimble libraries by Plataformatec:

  * [NimbleParsec](https://github.com/plataformatec/nimble_parsec) - simple and fast parser combinators
  * [NimbleCSV](https://github.com/plataformatec/nimble_csv) - simple and fast CSV parsing

# License

Copyright 2019 Plataformatec

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.md
## v0.1.1 (2019-10-15)

### Bug fixes

  * Drop default padding for %u to match other strftime implementations
  * Use `keyword()` type for options
  * Fix `-` modifier to actually drop padding per documentation

## v0.1.0 (2019-09-12)

  * First release


test/strftime_test.exs
defmodule StrftimeTest do
  use ExUnit.Case
  doctest Strftime

  describe "format/3" do
    test "return received string if there is no datetime formatting to be found in it" do
      assert Strftime.format(~N[2019-08-20 15:47:34.001], "muda string") == "muda string"
    end

    test "format all time zones blank when receiving a NaiveDateTime" do
      assert Strftime.format(~N[2019-08-15 17:07:57.001], "%z%Z") == ""
    end

    test "raise error when trying to format a date with a map that has no date fields" do
      time_without_date = %{hour: 15, minute: 47, second: 34, microsecond: {0, 0}}

      assert_raise(KeyError, fn -> Strftime.format(time_without_date, "%x") end)
    end

    test "raise error when trying to format a time with a map that has no time fields" do
      date_without_time = %{year: 2019, month: 8, day: 20}

      assert_raise(KeyError, fn -> Strftime.format(date_without_time, "%X") end)
    end

    test "raise error when the format is invalid" do
      assert_raise(FunctionClauseError, fn ->
        Strftime.format(~N[2019-08-20 15:47:34.001], "%-2-ç")
      end)
    end

    test "raise error when the preferred_datetime calls itself" do
      assert_raise(RuntimeError, fn ->
        Strftime.format(~N[2019-08-20 15:47:34.001], "%c", preferred_datetime: "%c")
      end)
    end

    test "raise error when the preferred_date calls itself" do
      assert_raise(RuntimeError, fn ->
        Strftime.format(~N[2019-08-20 15:47:34.001], "%x", preferred_date: "%x")
      end)
    end

    test "raise error when the preferred_time calls itself" do
      assert_raise(RuntimeError, fn ->
        Strftime.format(~N[2019-08-20 15:47:34.001], "%X", preferred_time: "%X")
      end)
    end

    test "raise error when the preferred formats create a circular chain" do
      assert_raise(RuntimeError, fn ->
        Strftime.format(~N[2019-08-20 15:47:34.001], "%c",
          preferred_datetime: "%x",
          preferred_date: "%X",
          preferred_time: "%c"
        )
      end)
    end

    test "format with no errors is the preferred formats are included multiple times on the same string" do
      assert(
        Strftime.format(~N[2019-08-15 17:07:57.001], "%c %c %x %x %X %X") ==
          "2019-08-15 17:07:57 2019-08-15 17:07:57 2019-08-15 2019-08-15 17:07:57 17:07:57"
      )
    end

    test "return `hour:minute:seconds PM` when receiving `%I:%M:%S %p`" do
      assert Strftime.format(~U[2019-08-15 17:07:57.001Z], "%I:%M:%S %p") == "05:07:57 PM"
    end

    test "ignore width when receiving the `-` padding option" do
      assert Strftime.format(~T[17:07:57.001], "%-999M") == "7"
    end

    test "format time zones correctly when receiving a DateTime" do
      datetime_with_zone = %DateTime{
        year: 2019,
        month: 8,
        day: 15,
        zone_abbr: "EEST",
        hour: 17,
        minute: 7,
        second: 57,
        microsecond: {0, 0},
        utc_offset: 7200,
        std_offset: 3600,
        time_zone: "UK"
      }

      assert Strftime.format(datetime_with_zone, "%z %Z") == "+0300 EEST"
    end

    test "return the formatted datetime when all format options and modifiers are received" do
      assert Strftime.format(
               ~U[2019-08-15 17:07:57.001Z],
               "%04% %a %A %b %B %-3c %d %f %H %I %j %m %_5M %p %P %q %S %u %x %X %y %Y %z %Z"
             ) ==
               "000% Thu Thursday Aug August 2019-08-15 17:07:57 15 1000 17 05 227 08     7 PM pm 3 57 04 2019-08-15 17:07:57 19 2019 +0000 UTC"
    end

    test "format according to received custom configs" do
      assert Strftime.format(
               ~U[2019-08-15 17:07:57.001Z],
               "%A %p %B %c %x %X",
               am_pm_names: {"a", "p"},
               month_names:
                 ~w(Janeiro Fevereiro Março Abril Maio Junho Julho Agosto Setembro Outubro Novembro Dezembro),
               day_of_week_names:
                 ~w(понедельник вторник среда четверг пятница суббота воскресенье),
               preferred_date: "%05Y-%m-%d",
               preferred_time: "%M:%_3H%S",
               preferred_datetime: "%%"
             ) == "четверг P Agosto % 02019-08-15 07: 1757"
    end
  end
end


test/strftime/format_stream_test.exs
defmodule Strftime.FormatStreamTest do
  use ExUnit.Case
  alias Strftime.FormatStream

  describe "stream/2" do
    test "return {received_format_stream, \"\"} when receiving an empty string" do
      format_stream = %FormatStream{format: nil, width: 5, pad: "_", section: "%_5"}
      assert FormatStream.stream("", format_stream) == {format_stream, ""}
    end

    test "return {updated_format_stream, \"\"} when receiving a valid stream from start to end" do
      expected_stream = %FormatStream{format: "A", width: 7, pad: "0", section: "%07A"}
      assert FormatStream.stream("07A", %FormatStream{}) == {expected_stream, ""}
    end

    test "keep format nil when the string ends too soon" do
      expected_stream = %FormatStream{format: nil, width: 7, pad: "0", section: "%07"}
      assert FormatStream.stream("07", %FormatStream{}) == {expected_stream, ""}
    end

    test "return the rest of the string after the stream ends" do
      assert {_stream, "-streamover"} = FormatStream.stream("07A-streamover", %FormatStream{})
    end
  end
end


mix.exs
defmodule NimbleStrftime.MixProject do
  use Mix.Project
  @version "0.1.1"
  @source_url "https://github.com/plataformatec/nimble_strftime"

  def project do
    [
      app: :nimble_strftime,
      version: @version,
      elixir: "~> 1.9",
      name: "NimbleStrftime",
      description: "strftime-based datetime formatter for Elixir",
      deps: deps(),
      docs: docs(),
      package: package()
    ]
  end

  defp docs do
    [
      main: "NimbleStrftime",
      source_ref: "v#{@version}",
      source_url: @source_url
    ]
  end

  defp package do
    %{
      licenses: ["Apache-2.0"],
      maintainers: ["Gustavo Santos Ferreira", "José Valim"],
      links: %{"GitHub" => @source_url}
    }
  end

  defp deps do
    [
      {:dialyxir, "~> 0.5", only: :dev},
      {:ex_doc, "~> 0.21", only: :dev},
      {:benchee, "~> 1.0", only: :dev}
    ]
  end
end


lib/strftime/format_stream.ex
defmodule Strftime.FormatStream do
  @moduledoc """
  Module that represents and prepares a Strftime `stream`.

  # What is a 'stream'?
    A stream, in this case is a sequence of bytes that may represent a
  date formatting instruction, the term was borrowed from the concept of
  data streams that work in the same principles as the format stream does

  # Format stream specification
    A stream here is made of up to 4 sections of which only the
  width has variable length
  stream sections
  <style>
    .stream-diagram {
      border: 1px solid;
      display: inline-block
    }
    .diagram-section {
      padding: 0em 1em;
      border-left: 1px solid;
      border-right: 1px solid;
      font-weight: 900;
    }
    .section-bytes {
      font-weight: 600;
    }
  </style>
  <div class="stream-diagram">
    <span class="diagram-section">
      init<span class="section-bytes">::1 byte</span>
    </span>
    <span class="diagram-section">
      pad<span class="section-bytes">::0 to 1byte</span>
    </span>
    <span class="diagram-section">
      width<span class="section-bytes">::0 to n bytes </span>
    </span>
    <span class="diagram-section">
      finalizer<span class="section-bytes">::1 byte or string end</span>
    </span>
  </div>
  * `init`: always the `%` char it signals the start of a section of formatted date
  * `pad`: can be one of the following charaters `0`, `-`, `_` it can also be empty
  * `width`: a sequence of numeric chars, can only start with zero if a pad was received in the previous section
  * `finalizer`: a non-numeric character that represents the formatting option, if the string ends short the format will be considered an empty string and the stream will end

  """

  @typedoc """
  A parsed(or partially parsed) format stream it holds the instructions to apply
  the date formatting once the stream was fully traversed
  * `format`: the final character of the stream that decides the formatting of the datetime
  * `width`: length the formatted stream will yield, some formats have default widths
  * `pad`: character that will be used on padding the result to the desired width, some formats have default pads
    * the `-` pad is a special case as it indicates the result must have no padding at all, nullifies the width
  * `section`: holds the parsed string as a fallback in case the stream holds invalid formatting instructions
  """
  @type t :: %__MODULE__{
          format: nil | String.t(),
          width: nil | integer(),
          pad: nil | String.t(),
          section: list(String.t())
        }
  defstruct [:format, :width, :pad, section: ["%"]]
end


lib/nimble_strftime/format_options.ex
defmodule NimbleStrftime.FormatOptions do
  @moduledoc """
  Module for setting and retrieving configurable formatting options
  """
  alias NimbleStrftime.FormatOptions

  @typedoc """
  Struct that holds configuration options, loads the application defaults on the configs
  """
  @type t :: %__MODULE__{
          preferred_date: String.t(),
          preferred_time: String.t(),
          preferred_datetime: String.t(),
          am_pm_names: {String.t(), String.t()},
          month_names: list(String.t()),
          day_of_week_names: list(String.t()),
          abbreviation_size: integer(),
          preferred_datetime_invoked: true | false,
          preferred_date_invoked: true | false,
          preferred_time_invoked: true | false
        }
  defstruct preferred_date: "%Y-%m-%d",
            preferred_time: "%H:%M:%S",
            preferred_datetime: "%Y-%m-%d %H:%M:%S",
            am_pm_names: {"am", "pm"},
            month_names:
              ~w(January February March April May June July August September October November December),
            day_of_week_names: ~w(Monday Tuesday Wednesday Thursday Friday Saturday Sunday),
            abbreviation_size: 3,
            preferred_datetime_invoked: false,
            preferred_date_invoked: false,
            preferred_time_invoked: false

  @typedoc "Accepted configuration options to be used by NimbleStrftime"
  @type options :: [option]

  @typedoc "Accepted configuration option to be used by NimbleStrftime"
  @type option ::
          {:preferred_date, String.t()}
          | {:preferred_time, String.t()}
          | {:preferred_datetime, String.t()}
          | {:am_pm_names, {String.t(), String.t()}}
          | {:month_names, list(String.t())}
          | {:day_of_week_names, list(String.t())}
          | {:abbreviation_size, integer()}

  @doc "Returns the `am_name` configured for the received format_options struct"
  @spec am_name(t()) :: String.t()
  def am_name(format_options = %FormatOptions{}), do: elem(format_options.am_pm_names, 0)

  @doc "Returns the `pm_name` configured for the received format_options struct"
  @spec pm_name(t()) :: String.t()
  def pm_name(format_options = %FormatOptions{}), do: elem(format_options.am_pm_names, 1)

  @doc "Returns the name of the received month number, based on the received format_options struct"
  @spec month_name(integer(), t()) :: String.t()
  def month_name(index, format_options = %FormatOptions{}) when index > 0 and index < 13 do
    Enum.at(format_options.month_names, index - 1)
  end

  @doc "Same as `month_name` but shortened to the configured `:abbreviation_size`"
  @spec month_name_abbreviated(integer(), t()) :: String.t()
  def month_name_abbreviated(index, format_options = %FormatOptions{}) do
    String.slice(month_name(index, format_options), 0..(format_options.abbreviation_size - 1))
  end

  @doc "Returns the name of the received week day number, based on the received options struct"
  @spec day_of_week_name(integer(), t()) :: String.t()
  def day_of_week_name(index, format_options = %FormatOptions{}) when index > 0 and index < 8 do
    Enum.at(format_options.day_of_week_names, index - 1)
  end

  @doc "Same as `day_of_week_name` but shortened to the configured `:abbreviation_size`"
  @spec day_of_week_name_abbreviated(integer(), t()) :: String.t()
  def day_of_week_name_abbreviated(index, format_options = %FormatOptions{}) do
    String.slice(
      day_of_week_name(index, format_options),
      0..(format_options.abbreviation_size - 1)
    )
  end
end


lib/nimble_strftime.ex
defmodule NimbleStrftime do
  @moduledoc """
  Simple datetime formatting based on the strftime format
  found on UNIX-like systems.

  ## Formatting syntax

  The formatting syntax for strftime is a sequence of characters in the following format:

      %<padding><width><format>

  where:

    * `%`: indicates the start of a formatted section
    * `<padding>`: set the padding (see below)
    * `<width>`: a number indicating the minimum size of the formatted section
    * `<format>`: the format iself (see below)

  ### Accepted padding options

    * `-`: no padding, removes all padding from the format
    * `_`: pad with spaces
    * `0`: pad with zeroes

  ### Accepted formats

  The accepted formats are:

  Format | Description                                                             | Examples (in ISO)
  :----- | :-----------------------------------------------------------------------| :------------------------
  a      | Abbreviated name of day                                                 | Mon
  A      | Full name of day                                                        | Monday
  b      | Abbreviated month name                                                  | Jan
  B      | Full month name                                                         | January
  c      | Preferred date+time representation                                      | 2018-10-17 12:34:56
  d      | Day of the month                                                        | 01, 12
  f      | Microseconds *(does not support width and padding modifiers)*           | 000000, 999999, 0123
  H      | Hour using a 24-hour clock                                              | 00, 23
  I      | Hour using a 12-hour clock                                              | 01, 12
  j      | Day of the year                                                         | 001, 366
  m      | Month                                                                   | 01, 12
  M      | Minute                                                                  | 00, 59
  p      | "AM" or "PM" (noon is "PM", midnight as "AM")                           | AM, PM
  P      | "am" or "pm" (noon is "pm", midnight as "am")                           | am, pm
  q      | Quarter                                                                 | 1, 2, 3, 4
  S      | Second                                                                  | 00, 59, 60
  u      | Day of the week                                                         | 1 (Monday), 7 (Sunday)
  x      | Preferred date (without time) representation                            | 2018-10-17
  X      | Preferred time (without date) representation                            | 12:34:56
  y      | Year as 2-digits                                                        | 01, 01, 86, 18
  Y      | Year                                                                    | -0001, 0001, 1986
  z      | +hhmm/-hhmm time zone offset from UTC (empty string if naive)           | +0300, -0530
  Z      | Time zone abbreviation (empty string if naive)                          | CET, BRST
  %      | Literal "%" character                                                   | %

  Any other character will be interpreted as an invalid format and raise an error
  """

  @doc """
  Formats received datetime into a string.

  The datetime can be any of the Calendar types (`Time`, `Date`,
  `NaiveDateTime`, and `DateTime`) or any map, as long as they
  contain all of the relevant fields necessary for formatting.
  For example, if you use `%Y` to format the year, the datatime
  must have the `:year` field. Therefore, if you pass a `Time`,
  or a map without the `:year` field to a format that expects `%Y`,
  an error will be raised.

  ## Options

    * `:preferred_datetime` - a string for the preferred format to show datetimes,
      it can't contain the `%c` format and defaults to `"%Y-%m-%d %H:%M:%S"`
      if the option is not received

    * `:preferred_date` - a string for the preferred format to show dates,
      it can't contain the `%x` format and defaults to `"%Y-%m-%d"`
      if the option is not received

    * `:preferred_time` - a string for the preferred format to show times,
      it can't contain the `%X` format and defaults to `"%H:%M:%S"`
      if the option is not received

    * `:am_pm_names` - a function that receives either `:am` or `:pm` and returns
      the name of the period of the day, if the option is not received it defaults
      to a function that returns `"am"` and `"pm"`, respectively

    *  `:month_names` - a function that receives a number and returns the name of
      the corresponding month, if the option is not received it defaults to a
      function thet returns the month names in english

    * `:abbreviated_month_names` - a function that receives a number and returns the
      abbreviated name of the corresponding month, if the option is not received it
      defaults to a function thet returns the abbreviated month names in english

    * `:day_of_week_names` - a function that receives a number and returns the name of
      the corresponding day of week, if the option is not received it defaults to a
      function that returns the day of week names in english

    * `:abbreviated_day_of_week_names` - a function that receives a number and returns
      the abbreviated name of the corresponding day of week, if the option is not received 
      it defaults to a function that returns the abbreviated day of week names in english

  ## Examples

  Without options:

      iex> NimbleStrftime.format(~U[2019-08-26 13:52:06.0Z], "%y-%m-%d %I:%M:%S %p")
      "19-08-26 01:52:06 PM"

      iex> NimbleStrftime.format(~U[2019-08-26 13:52:06.0Z], "%a, %B %d %Y")
      "Mon, August 26 2019"

      iex> NimbleStrftime.format(~U[2019-08-26 13:52:06.0Z], "%c")
      "2019-08-26 13:52:06"

  With options:

      iex> NimbleStrftime.format(~U[2019-08-26 13:52:06.0Z], "%c", preferred_datetime: "%H:%M:%S %d-%m-%y")
      "13:52:06 26-08-19"

      iex> NimbleStrftime.format(
      ...>  ~U[2019-08-26 13:52:06.0Z],
      ...>  "%A",
      ...>  day_of_week_names: fn day_of_week ->
      ...>    {"segunda-feira", "terça-feira", "quarta-feira", "quinta-feira",
      ...>    "sexta-feira", "sábado", "domingo"}
      ...>    |> elem(day_of_week - 1)
      ...>  end
      ...>)
      "segunda-feira"

      iex> NimbleStrftime.format(
      ...>  ~U[2019-08-26 13:52:06.0Z],
      ...>  "%B",
      ...>  month_names: fn month ->
      ...>    {"январь", "февраль", "март", "апрель", "май", "июнь",
      ...>    "июль", "август", "сентябрь", "октябрь", "ноябрь", "декабрь"}
      ...>    |> elem(month - 1)
      ...>  end
      ...>)
      "август"
  """
  @spec format(map(), String.t(), keyword()) :: String.t()
  def format(date_or_time_or_datetime, string_format, user_options \\ []) do
    parse(
      string_format,
      date_or_time_or_datetime,
      options(user_options),
      []
    )
    |> IO.iodata_to_binary()
  end

  defp parse("", _datetime, _format_options, acc),
    do: Enum.reverse(acc)

  defp parse("%" <> rest, datetime, format_options, acc),
    do: parse_modifiers(rest, nil, nil, {datetime, format_options, acc})

  defp parse(<<char::binary-1, rest::binary>>, datetime, format_options, acc),
    do: parse(rest, datetime, format_options, [char | acc])

  defp parse_modifiers("-" <> rest, width, nil, parser_data) do
    parse_modifiers(rest, width, "", parser_data)
  end

  defp parse_modifiers("0" <> rest, width, nil, parser_data) do
    parse_modifiers(rest, width, ?0, parser_data)
  end

  defp parse_modifiers("_" <> rest, width, nil, parser_data) do
    parse_modifiers(rest, width, ?\s, parser_data)
  end

  defp parse_modifiers(<<digit, rest::binary>>, width, pad, parser_data) when digit in ?0..?9 do
    new_width =
      case pad do
        ?- -> 0
        _ -> (width || 0) * 10 + (digit - ?0)
      end

    parse_modifiers(rest, new_width, pad, parser_data)
  end

  # set default padding if none was specfied
  defp parse_modifiers(<<format, _::binary>> = rest, width, nil, parser_data) do
    parse_modifiers(rest, width, default_pad(format), parser_data)
  end

  # set default width if none was specified
  defp parse_modifiers(<<format, _::binary>> = rest, nil, pad, parser_data) do
    parse_modifiers(rest, default_width(format), pad, parser_data)
  end

  defp parse_modifiers(rest, width, pad, {datetime, format_options, acc}) do
    format_modifiers(rest, width, pad, datetime, format_options, acc)
  end

  defp am_pm(hour, format_options) when hour > 11 do
    format_options.am_pm_names.(:pm)
  end

  defp am_pm(hour, format_options) when hour <= 11 do
    format_options.am_pm_names.(:am)
  end

  defp default_pad(format) when format in 'aAbBpPZ', do: ?\s
  defp default_pad(_format), do: ?0

  defp default_width(format) when format in 'dHImMSy', do: 2
  defp default_width(?j), do: 3
  defp default_width(format) when format in 'Yz', do: 4
  defp default_width(_format), do: 0

  # Literally just %
  defp format_modifiers("%" <> rest, width, pad, datetime, format_options, acc) do
    parse(rest, datetime, format_options, [pad_leading("%", width, pad) | acc])
  end

  # Abbreviated name of day
  defp format_modifiers("a" <> rest, width, pad, datetime, format_options, acc) do
    result =
      datetime
      |> Date.day_of_week()
      |> format_options.abbreviated_day_of_week_names.()
      |> pad_leading(width, pad)

    parse(rest, datetime, format_options, [result | acc])
  end

  # Full name of day
  defp format_modifiers("A" <> rest, width, pad, datetime, format_options, acc) do
    result =
      datetime
      |> Date.day_of_week()
      |> format_options.day_of_week_names.()
      |> pad_leading(width, pad)

    parse(rest, datetime, format_options, [result | acc])
  end

  # Abbreviated month name
  defp format_modifiers("b" <> rest, width, pad, datetime, format_options, acc) do
    result =
      datetime.month
      |> format_options.abbreviated_month_names.()
      |> pad_leading(width, pad)

    parse(rest, datetime, format_options, [result | acc])
  end

  # Full month name
  defp format_modifiers("B" <> rest, width, pad, datetime, format_options, acc) do
    result = datetime.month |> format_options.month_names.() |> pad_leading(width, pad)

    parse(rest, datetime, format_options, [result | acc])
  end

  # Preferred date+time representation
  defp format_modifiers(
         "c" <> _rest,
         _width,
         _pad,
         _datetime,
         %{preferred_datetime_invoked: true},
         _acc
       ) do
    raise RuntimeError,
          "tried to format preferred_datetime within another preferred_datetime format"
  end

  defp format_modifiers("c" <> rest, width, pad, datetime, format_options, acc) do
    result =
      format_options.preferred_datetime
      |> parse(datetime, %{format_options | preferred_datetime_invoked: true}, [])
      |> pad_preferred(width, pad)

    parse(rest, datetime, format_options, [result | acc])
  end

  # Day of the month
  defp format_modifiers("d" <> rest, width, pad, datetime, format_options, acc) do
    result = datetime.day |> Integer.to_string() |> pad_leading(width, pad)
    parse(rest, datetime, format_options, [result | acc])
  end

  # Microseconds
  defp format_modifiers("f" <> rest, _width, _pad, datetime, format_options, acc) do
    {microsecond, precision} = datetime.microsecond

    result =
      microsecond
      |> Integer.to_string()
      |> String.pad_leading(6, "0")
      |> binary_part(0, max(precision, 1))

    parse(rest, datetime, format_options, [result | acc])
  end

  # Hour using a 24-hour clock
  defp format_modifiers("H" <> rest, width, pad, datetime, format_options, acc) do
    result = datetime.hour |> Integer.to_string() |> pad_leading(width, pad)
    parse(rest, datetime, format_options, [result | acc])
  end

  # Hour using a 12-hour clock
  defp format_modifiers("I" <> rest, width, pad, datetime, format_options, acc) do
    result = (rem(datetime.hour() + 23, 12) + 1) |> Integer.to_string() |> pad_leading(width, pad)
    parse(rest, datetime, format_options, [result | acc])
  end

  # Day of the year
  defp format_modifiers("j" <> rest, width, pad, datetime, format_options, acc) do
    result = datetime |> Date.day_of_year() |> Integer.to_string() |> pad_leading(width, pad)
    parse(rest, datetime, format_options, [result | acc])
  end

  # Month
  defp format_modifiers("m" <> rest, width, pad, datetime, format_options, acc) do
    result = datetime.month |> Integer.to_string() |> pad_leading(width, pad)
    parse(rest, datetime, format_options, [result | acc])
  end

  # Minute
  defp format_modifiers("M" <> rest, width, pad, datetime, format_options, acc) do
    result = datetime.minute |> Integer.to_string() |> pad_leading(width, pad)
    parse(rest, datetime, format_options, [result | acc])
  end

  # “AM” or “PM” (noon is “PM”, midnight as “AM”)
  defp format_modifiers("p" <> rest, width, pad, datetime, format_options, acc) do
    result = datetime.hour |> am_pm(format_options) |> String.upcase() |> pad_leading(width, pad)

    parse(rest, datetime, format_options, [result | acc])
  end

  # “am” or “pm” (noon is “pm”, midnight as “am”)
  defp format_modifiers("P" <> rest, width, pad, datetime, format_options, acc) do
    result =
      datetime.hour
      |> am_pm(format_options)
      |> String.downcase()
      |> pad_leading(width, pad)

    parse(rest, datetime, format_options, [result | acc])
  end

  # Quarter
  defp format_modifiers("q" <> rest, width, pad, datetime, format_options, acc) do
    result = datetime |> Date.quarter_of_year() |> Integer.to_string() |> pad_leading(width, pad)
    parse(rest, datetime, format_options, [result | acc])
  end

  # Second
  defp format_modifiers("S" <> rest, width, pad, datetime, format_options, acc) do
    result = datetime.second |> Integer.to_string() |> pad_leading(width, pad)
    parse(rest, datetime, format_options, [result | acc])
  end

  # Day of the week
  defp format_modifiers("u" <> rest, width, pad, datetime, format_options, acc) do
    result = datetime |> Date.day_of_week() |> Integer.to_string() |> pad_leading(width, pad)
    parse(rest, datetime, format_options, [result | acc])
  end

  # Preferred date (without time) representation
  defp format_modifiers(
         "x" <> _rest,
         _width,
         _pad,
         _datetime,
         %{preferred_date_invoked: true},
         _acc
       ) do
    raise RuntimeError,
          "tried to format preferred_date within another preferred_date format"
  end

  defp format_modifiers("x" <> rest, width, pad, datetime, format_options, acc) do
    result =
      format_options.preferred_date
      |> parse(datetime, %{format_options | preferred_date_invoked: true}, [])
      |> pad_preferred(width, pad)

    parse(rest, datetime, format_options, [result | acc])
  end

  # Preferred time (without date) representation
  defp format_modifiers(
         "X" <> _rest,
         _width,
         _pad,
         _datetime,
         %{preferred_time_invoked: true},
         _acc
       ) do
    raise RuntimeError,
          "tried to format preferred_time within another preferred_time format"
  end

  defp format_modifiers("X" <> rest, width, pad, datetime, format_options, acc) do
    result =
      format_options.preferred_time
      |> parse(datetime, %{format_options | preferred_time_invoked: true}, [])
      |> pad_preferred(width, pad)

    parse(rest, datetime, format_options, [result | acc])
  end

  # Year as 2-digits
  defp format_modifiers("y" <> rest, width, pad, datetime, format_options, acc) do
    result = datetime.year |> rem(100) |> Integer.to_string() |> pad_leading(width, pad)
    parse(rest, datetime, format_options, [result | acc])
  end

  # Year
  defp format_modifiers("Y" <> rest, width, pad, datetime, format_options, acc) do
    result = datetime.year |> Integer.to_string() |> pad_leading(width, pad)
    parse(rest, datetime, format_options, [result | acc])
  end

  # +hhmm/-hhmm time zone offset from UTC (empty string if naive)
  defp format_modifiers(
         "z" <> rest,
         width,
         pad,
         datetime = %{utc_offset: utc_offset, std_offset: std_offset},
         format_options,
         acc
       ) do
    absolute_offset = abs(utc_offset + std_offset)

    offset_number =
      Integer.to_string(div(absolute_offset, 3600) * 100 + rem(div(absolute_offset, 60), 60))

    sign = if utc_offset + std_offset >= 0, do: "+", else: "-"
    result = "#{sign}#{pad_leading(offset_number, width, pad)}"
    parse(rest, datetime, format_options, [result | acc])
  end

  defp format_modifiers("z" <> rest, _width, _pad, datetime, format_options, acc) do
    parse(rest, datetime, format_options, ["" | acc])
  end

  # Time zone abbreviation (empty string if naive)
  defp format_modifiers("Z" <> rest, width, pad, datetime, format_options, acc) do
    result = datetime |> Map.get(:zone_abbr, "") |> pad_leading(width, pad)
    parse(rest, datetime, format_options, [result | acc])
  end

  defp pad_preferred(result, width, pad) when length(result) < width do
    pad_preferred([pad | result], width, pad)
  end

  defp pad_preferred(result, _width, _pad), do: result

  defp pad_leading(string, count, padding) do
    to_pad = count - byte_size(string)
    if to_pad > 0, do: do_pad_leading(to_pad, padding, string), else: string
  end

  defp do_pad_leading(0, _, acc), do: acc

  defp do_pad_leading(count, padding, acc),
    do: do_pad_leading(count - 1, padding, [padding | acc])

  defp options(user_options) do
    default_options = %{
      preferred_date: "%Y-%m-%d",
      preferred_time: "%H:%M:%S",
      preferred_datetime: "%Y-%m-%d %H:%M:%S",
      am_pm_names: fn
        :am -> "am"
        :pm -> "pm"
      end,
      month_names: fn month ->
        {"January", "February", "March", "April", "May", "June", "July", "August", "September",
         "October", "November", "December"}
        |> elem(month - 1)
      end,
      day_of_week_names: fn day_of_week ->
        {"Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"}
        |> elem(day_of_week - 1)
      end,
      abbreviated_month_names: fn month ->
        {"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"}
        |> elem(month - 1)
      end,
      abbreviated_day_of_week_names: fn day_of_week ->
        {"Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"} |> elem(day_of_week - 1)
      end,
      preferred_datetime_invoked: false,
      preferred_date_invoked: false,
      preferred_time_invoked: false
    }

    Enum.reduce(user_options, default_options, fn {key, value}, acc ->
      if Map.has_key?(acc, key) do
        %{acc | key => value}
      else
        acc
      end
    end)
  end
end
